### PR TITLE
Widget tests

### PR DESCRIPTION
**Pull request by blink1073**
Adds tests for `widget.ts`

```
----------------|----------|----------|----------|----------|----------------|
File            |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
----------------|----------|----------|----------|----------|----------------|
  widget.js     |    99.22 |    90.07 |      100 |    99.73 |            892 |
```

### CODE REVIEW

File: test/src/ui/widget.spec.ts [hunk 1]
```diff
@@ -0,0 +1,1792 @@
...
+import {
+  Message, sendMessage
+} from '../../../lib/core/messaging';
+
+import {
+  Title
+} from '../../../lib/ui/title';
+
+import {
+  Layout, Widget, WidgetFlag, ChildMessage, ResizeMessage, WidgetMessage
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: ascii sort order please

File: test/src/ui/widget.spec.ts [hunk 2]
```diff
@@ -0,0 +1,1792 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) 2014-2016, PhosphorJS Contributors
+|
+| Distributed under the terms of the BSD 3-Clause License.
+|
+| The full license is in the file LICENSE, distributed with this software.
+|----------------------------------------------------------------------------*/
+'use strict';
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: use strict is no longer required as of TS 1.8 - it gets added automatically by the compiler

File: test/src/ui/widget.spec.ts [hunk 3]
```diff
@@ -0,0 +1,1792 @@
...
+        Widget.detach(widget);
+        expect(widget.messages[0]).to.be('before-detach');
+        widget.dispose();
+      });
+
+    });
+
+    describe('.prepareGeometry()', () => {
+
+      it('should se the inline style position of the widget to `absolute`', () => {
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: s/se/set

File: test/src/ui/widget.spec.ts [hunk 4]
```diff
@@ -0,0 +1,1792 @@
...
+      it('should be `true` if the widget is disposed', () => {
+        let widget = new Widget();
+        widget.dispose();
+        expect(widget.isDisposed).to.be(true);
+      });
+
+      it('should be `false` if the widget is not disposed', () => {
+        let widget = new Widget();
+        expect(widget.isDisposed).to.be(false);
+      });
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: need a test for read-only

File: test/src/ui/widget.spec.ts [hunk 5]
```diff
@@ -0,0 +1,1792 @@
...
+        let widget = new Widget();
+        Widget.attach(widget, document.body);
+        expect(widget.isAttached).to.be(true);
+        widget.dispose();
+      });
+
+      it('should be `false` if the widget is not attached', () => {
+        let widget = new Widget();
+        expect(widget.isAttached).to.be(false);
+      });
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: need a test for read-only

File: test/src/ui/widget.spec.ts [hunk 6]
```diff
@@ -0,0 +1,1792 @@
...
+        expect(widget.isHidden).to.be(true);
+        widget.dispose();
+      });
+
+      it('should be `false` if the widget is not hidden', () => {
+        let widget = new Widget();
+        Widget.attach(widget, document.body);
+        expect(widget.isHidden).to.be(false);
+        widget.dispose();
+      });
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: need a test for read-only

File: test/src/ui/widget.spec.ts [hunk 7]
```diff
@@ -0,0 +1,1792 @@
...
+        Widget.attach(widget, document.body);
+        widget.hide();
+        expect(widget.isVisible).to.be(false);
+        widget.dispose();
+      });
+
+      it('should be `false` if the widget is not attached', () => {
+        let widget = new Widget();
+        expect(widget.isVisible).to.be(false);
+      });
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: need a test for read-only

File: test/src/ui/widget.spec.ts [hunk 8]
```diff
@@ -0,0 +1,1792 @@
...
+      it("should return an iterator over the widget's children", () => {
+        let widget = new Widget();
+        widget.layout = new LogLayout();
+        let child = widget.children().next();
+        expect(child instanceof LogWidget).to.be(true);
+      });
+
+      it('should return an empty iterator if there is no layout', () => {
+        let widget = new Widget();
+        expect(widget.children()).to.be(EmptyIterator.instance);
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: this is testing an implementation detail. Better to exhaust the iterator into an array, and test that the array length is zero:

`expect(toArray(widget.children()).to.eql([])`

File: test/src/ui/widget.spec.ts [hunk 9]
```diff
@@ -0,0 +1,1792 @@
...
+        widget.setFlag(WidgetFlag.IsHidden);
+        widget.clearFlag(WidgetFlag.IsHidden);
+        expect(widget.testFlag(WidgetFlag.IsHidden)).to.be(false);
+      });
+
+    });
+
+    describe('#notifyLayout()', () => {
+
+      it("should invoke the message processing routine of the widget's layout", () => {
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: this description is weird. it's not actually what the test is testing.

File: test/src/ui/widget.spec.ts [hunk 10]
```diff
@@ -0,0 +1,1792 @@
...
+      });
+
+      it('should be called automatically when the parent is disposed', () => {
+        let widget = new Widget();
+        let layout = new LogLayout();
+        widget.layout = layout;
+        widget.dispose();
+        expect(layout.parent).to.be(null);
+        expect(layout.isDisposed).to.be(true);
+      });
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: it should also dispose of all child widgets, which means your custom layout needs to reimplement `dispose` to do that.

File: test/src/ui/widget.spec.ts [hunk 11]
```diff
@@ -0,0 +1,1792 @@
...
+        expect(layout.methods.indexOf('onResize')).to.not.be(-1);
+      });
+
+      it('should send a `resize` message to each of the widgets in the layout', () => {
+        let layout = new LogLayout();
+        let parent = new Widget();
+        parent.layout = layout;
+        sendMessage(parent, ResizeMessage.UnknownSize);
+        expect(layout.methods.indexOf('onResize')).to.not.be(-1);
+        let child = layout.iter().next();
```
Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: you should test each child, not just the first. The `iteration` module has helpers for this:

`expect(every(layout, child => child.methods.indexOf('onResize') !== -1)).to.be(true)`

Comment (sccolbert) [inline test/src/ui/widget.spec.ts]: this should be fixed for every test verifying iteration over layout children.

**Comment by blink1073** [comment]
Comments addressed.

**Pull request merged**